File: src/main/java/io/mateo/cxf/codegen/wsdl2java/WsdlOption.java
/*
 * Copyright 2020 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.mateo.cxf.codegen.wsdl2java;

import org.gradle.api.Named;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.ProjectLayout;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.ListProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.SetProperty;

import javax.inject.Inject;
import java.util.List;

/**
 * Default implementation for options for the {@code wsdl2java} command.
 */
public class WsdlOption implements Option, Named {

	private final String name;

	private final RegularFileProperty wsdl;

	private final ListProperty<String> packageNames;

	private final ListProperty<String> extraArgs;

	private final ListProperty<String> xjcArgs;

	private final ListProperty<String> asyncMethods;

	private final ListProperty<String> bareMethods;

	private final ListProperty<String> mimeMethods;

	private final ListProperty<String> namespaceExcludes;

	private final Property<Boolean> defaultExcludesNamespace;

	private final Property<Boolean> defaultNamespacePackageMapping;

	private final SetProperty<String> bindingFiles;

	private final Property<String> wsdlLocation;

	private final Property<Boolean> wsdlList;

	private final Property<String> frontend;

	private final Property<String> databinding;

	private final Property<String> wsdlVersion;

	private final Property<String> catalog;

	private final Property<Boolean> extendedSoapHeaders;

	private final Property<String> validateWsdl;

	private final Property<Boolean> noTypes;

	private final Property<String> faultSerialVersionUid;

	private final Property<String> exceptionSuper;

	private final ListProperty<String> seiSuper;

	private final Property<Boolean> markGenerated;

	private final Property<Boolean> suppressGeneratedData;

	private final Property<String> serviceName;

	private final Property<Boolean> autoNameResolution;

	private final Property<Boolean> noAddressBinding;

	private final Property<Boolean> allowElementRefs;

	private final Property<Boolean> verbose;

	private final DirectoryProperty outputDir;

	@Inject
	public WsdlOption(String name, ObjectFactory objects, ProjectLayout layout) {
		this.name = name;
		this.outputDir = objects.directoryProperty().convention(layout.getBuildDirectory().dir("/generated-sources"));
		this.wsdl = objects.fileProperty();
		this.packageNames = objects.listProperty(String.class);
		this.extraArgs = objects.listProperty(String.class);
		this.xjcArgs = objects.listProperty(String.class);
		this.asyncMethods = objects.listProperty(String.class);
		this.bareMethods = objects.listProperty(String.class);
		this.mimeMethods = objects.listProperty(String.class);
		this.namespaceExcludes = objects.listProperty(String.class);
		this.defaultExcludesNamespace = objects.property(Boolean.class);
		this.defaultNamespacePackageMapping = objects.property(Boolean.class);
		this.bindingFiles = objects.setProperty(String.class);
		this.wsdlLocation = objects.property(String.class);
		this.wsdlList = objects.property(Boolean.class);
		this.frontend = objects.property(String.class);
		this.databinding = objects.property(String.class);
		this.wsdlVersion = objects.property(String.class);
		this.catalog = objects.property(String.class);
		this.extendedSoapHeaders = objects.property(Boolean.class);
		this.validateWsdl = objects.property(String.class);
		this.noTypes = objects.property(Boolean.class);
		this.faultSerialVersionUid = objects.property(String.class);
		this.exceptionSuper = objects.property(String.class);
		this.seiSuper = objects.listProperty(String.class);
		this.markGenerated = objects.property(Boolean.class);
		this.suppressGeneratedData = objects.property(Boolean.class);
		this.serviceName = objects.property(String.class);
		this.autoNameResolution = objects.property(Boolean.class);
		this.noAddressBinding = objects.property(Boolean.class);
		this.allowElementRefs = objects.property(Boolean.class);
		this.verbose = objects.property(Boolean.class);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public RegularFileProperty getWsdl() {
		return this.wsdl;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getPackageNames() {
		return this.packageNames;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getExtraArgs() {
		return this.extraArgs;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getXjcArgs() {
		return this.xjcArgs;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getAsyncMethods() {
		return this.asyncMethods;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getBareMethods() {
		return this.bareMethods;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getMimeMethods() {
		return this.mimeMethods;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getNamespaceExcludes() {
		return this.namespaceExcludes;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getDefaultExcludesNamespace() {
		return this.defaultExcludesNamespace;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getDefaultNamespacePackageMapping() {
		return this.defaultNamespacePackageMapping;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public SetProperty<String> getBindingFiles() {
		return this.bindingFiles;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getWsdlLocation() {
		return this.wsdlLocation;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getWsdlList() {
		return this.wsdlList;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getFrontend() {
		return this.frontend;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getDatabinding() {
		return this.databinding;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getWsdlVersion() {
		return this.wsdlVersion;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getCatalog() {
		return this.catalog;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getExtendedSoapHeaders() {
		return this.extendedSoapHeaders;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getValidateWsdl() {
		return this.validateWsdl;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getNoTypes() {
		return this.noTypes;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getFaultSerialVersionUid() {
		return this.faultSerialVersionUid;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getExceptionSuper() {
		return this.exceptionSuper;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ListProperty<String> getSeiSuper() {
		return this.seiSuper;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getMarkGenerated() {
		return this.markGenerated;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getSuppressGeneratedData() {
		return this.suppressGeneratedData;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<String> getServiceName() {
		return this.serviceName;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getAutoNameResolution() {
		return this.autoNameResolution;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getNoAddressBinding() {
		return this.noAddressBinding;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getAllowElementRefs() {
		return this.allowElementRefs;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Property<Boolean> getVerbose() {
		return this.verbose;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String getName() {
		return this.name;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public DirectoryProperty getOutputDir() {
		return this.outputDir;
	}

	public List<String> generateArgs() {

		return null;
	}
}

File: src/main/kotlin/io/mateo/cxf/codegen/wsdl2java/WsdlOption.kt
/*
 * Copyright 2020 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.mateo.cxf.codegen.wsdl2java

import javax.inject.Inject
import org.gradle.api.GradleException
import org.gradle.api.Named
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.ProjectLayout
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.Property
import org.gradle.api.provider.SetProperty

open class WsdlOption @Inject constructor(
    private val name: String,
    objects: ObjectFactory,
    private val layout: ProjectLayout
) : Option, Named {

    override val wsdl: RegularFileProperty = objects.fileProperty()

    override val packageNames: ListProperty<String> = objects.listProperty(String::class.java)

    override val extraArgs: ListProperty<String> = objects.listProperty(String::class.java)

    override val xjcArgs: ListProperty<String> = objects.listProperty(String::class.java)

    override val asyncMethods: ListProperty<String> = objects.listProperty(String::class.java)

    override val bareMethods: ListProperty<String> = objects.listProperty(String::class.java)

    override val mimeMethods: ListProperty<String> = objects.listProperty(String::class.java)

    override val namespaceExcludes: ListProperty<String> = objects.listProperty(String::class.java)

    override val defaultExcludesNamespace: Property<Boolean> = objects.property(Boolean::class.java)

    override val defaultNamespacePackageMapping: Property<Boolean> = objects.property(Boolean::class.java)

    override val bindingFiles: SetProperty<String> = objects.setProperty(String::class.java)

    override val wsdlLocation: Property<String> = objects.property(String::class.java)

    override val wsdlList: Property<Boolean> = objects.property(Boolean::class.java)

    override val frontend: Property<String> = objects.property(String::class.java)

    override val databinding: Property<String> = objects.property(String::class.java)

    override val wsdlVersion: Property<String> = objects.property(String::class.java)

    override val catalog: Property<String> = objects.property(String::class.java)

    override val extendedSoapHeaders: Property<Boolean> = objects.property(Boolean::class.java)

    override val validateWsdl: Property<String> = objects.property(String::class.java)

    override val noTypes: Property<Boolean> = objects.property(Boolean::class.java)

    override val faultSerialVersionUid: Property<String> = objects.property(String::class.java)

    override val exceptionSuper: Property<String> = objects.property(String::class.java)

    override val seiSuper: ListProperty<String> = objects.listProperty(String::class.java)

    override val markGenerated: Property<Boolean> = objects.property(Boolean::class.java)

    override val suppressGeneratedData: Property<Boolean> = objects.property(Boolean::class.java)

    override val serviceName: Property<String> = objects.property(String::class.java)

    override val autoNameResolution: Property<Boolean> = objects.property(Boolean::class.java)

    override val noAddressBinding: Property<Boolean> = objects.property(Boolean::class.java)

    override val allowElementRefs: Property<Boolean> = objects.property(Boolean::class.java)

    override val verbose: Property<Boolean> = objects.property(Boolean::class.java)

    override val outputDir: DirectoryProperty = objects.directoryProperty().convention(layout.buildDirectory.dir("/generated-sources"))

    override fun getName() = name

    override fun toString() = StringBuilder(128).apply {
        append("WSDL: ").append(wsdl.get().asFile.toURI()).append('\n')
        append("OutputDir: ").append(outputDir.get().asFile.absolutePath).append('\n')
        append("Extraargs: ").append(extraArgs.get()).append('\n').append('\n')
        append("XJCargs: ").append(xjcArgs.get()).append('\n')
        append("Packagenames: ").append(packageNames.get()).append('\n')
    }.toString()

    fun generateArgs(): List<String> {
        val command = mutableListOf<String>()
        if (packageNames.isPresent) {
            packageNames.get().forEach {
                command.add("-p")
                command.add(it)
            }
        }
        if (namespaceExcludes.isPresent) {
            namespaceExcludes.get().forEach {
                command.add("-nexclude")
                command.add(it)
            }
        }
        command.add("-d")
        command.add(outputDir.get().asFile.absolutePath) // always set by convention
        if (bindingFiles.isPresent) {
            bindingFiles.get().forEach {
                val bindingFile = layout.projectDirectory.file(it)
                command.add("-b")
                command.add(bindingFile.asFile.toURI().toString())
            }
        }
        if (frontend.isPresent) {
            command.add("-fe")
            command.add(frontend.get())
        }
        if (databinding.isPresent) {
            command.add("-db")
            command.add(databinding.get())
        }
        if (wsdlVersion.isPresent) {
            command.add("-wv")
            command.add(wsdlVersion.get())
        }
        if (catalog.isPresent) {
            command.add("-catalog")
            command.add(catalog.get())
        }
        if (extendedSoapHeaders.isPresent && extendedSoapHeaders.get()) {
            command.add("-exsh")
            command.add("true")
        }
        if (noTypes.isPresent && noTypes.get()) {
            command.add("-noTypes")
        }
        if (allowElementRefs.isPresent && allowElementRefs.get()) {
            command.add("-allowElementReferences")
        }
        if (validateWsdl.isPresent) {
            command.add("-validate=${validateWsdl.get()}")
        }
        if (markGenerated.isPresent && markGenerated.get()) {
            command.add("-mark-generated")
        }
        if (suppressGeneratedData.isPresent && suppressGeneratedData.get()) {
            command.add("-suppress-generated-date")
        }
        if (defaultExcludesNamespace.isPresent) {
            command.add("-dex")
            command.add(defaultExcludesNamespace.get().toString())
        }
        if (defaultNamespacePackageMapping.isPresent) {
            command.add("-dns")
            command.add(defaultNamespacePackageMapping.get().toString())
        }
        if (serviceName.isPresent) {
            command.add("-sn")
            command.add(serviceName.get())
        }
        if (faultSerialVersionUid.isPresent) {
            command.add("-faultSerialVersionUID")
            command.add(faultSerialVersionUid.get())
        }
        if (exceptionSuper.isPresent) {
            command.add("-exceptionSuper")
            command.add(exceptionSuper.get())
        }
        if (seiSuper.isPresent) {
            seiSuper.get().forEach {
                command.add("-seiSuper")
                command.add(it)
            }
        }
        if (autoNameResolution.isPresent && autoNameResolution.get()) {
            command.add("-autoNameResolution")
        }
        if (noAddressBinding.isPresent && noAddressBinding.get()) {
            command.add("-noAddressBinding")
        }
        if (xjcArgs.isPresent) {
            xjcArgs.get().forEach {
                command.add("-xjc$it")
            }
        }
        if (extraArgs.isPresent) {
            command.addAll(extraArgs.get())
        }
        if (wsdlLocation.isPresent) {
            command.add("-wsdlLocation")
            command.add(wsdlLocation.get())
        }
        if (wsdlList.isPresent && wsdlList.get()) {
            command.add("-wsdlList")
        }
        if (verbose.isPresent && verbose.get()) {
            command.add("-verbose")
        }
        if (asyncMethods.isPresent && asyncMethods.get().isNotEmpty()) {
            val sb = StringBuilder("-asyncMethods").apply {
                append("=")
                var first = true
                for (value in asyncMethods.get()) {
                    if (!first) {
                        append(",")
                    }
                    append(value)
                    first = false
                }
            }
            command.add(sb.toString())
        }
        if (bareMethods.isPresent && bareMethods.get().isNotEmpty()) {
            val sb = StringBuilder("-bareMethods").apply {
                append("=")
                var first = true
                for (value in bareMethods.get()) {
                    if (!first) {
                        append(",")
                    }
                    append(value)
                    first = false
                }
            }
            command.add(sb.toString())
        }
        if (mimeMethods.isPresent && mimeMethods.get().isNotEmpty()) {
            val sb = StringBuilder("-mimeMethods").apply {
                append("=")
                var first = true
                for (value in mimeMethods.get()) {
                    if (!first) {
                        append(",")
                    }
                    append(value)
                    first = false
                }
            }
            command.add(sb.toString())
        }
        if (!wsdl.isPresent) {
            throw GradleException("'wsdl' property is not present")
        }
        command.add(wsdl.get().asFile.toURI().toString())
        return command
    }
}
